crates/pathfinder/src/rpc/v02/method/get_block.rs

```diff
@@ -0,0 +1,414 @@
+use anyhow::{anyhow, Context};
+use serde::Deserialize;
+use stark_hash::StarkHash;
+
+use crate::core::{BlockId, GlobalRoot, StarknetBlockHash, StarknetBlockNumber};
+use crate::rpc::v02::common::get_block_status;
+use crate::rpc::v02::RpcContext;
+use crate::storage::{StarknetBlocksBlockId, StarknetBlocksTable, StarknetTransactionsTable};
+
+#[derive(Deserialize, Debug, PartialEq, Eq)]
+#[cfg_attr(test, derive(Copy, Clone))]
+pub struct GetBlockInput {
+    block_id: BlockId,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(GetBlockError: BlockNotFound);
+
+/// Get block information with transaction hashes given the block id
+pub async fn get_block_with_tx_hashes(
+    context: RpcContext,
+    input: GetBlockInput,
+) -> Result<types::Block, GetBlockError> {
+    get_block(
+        context,
+        input.block_id,
+        types::BlockResponseScope::TransactionHashes,
+    )
+    .await
+}
+
+/// Get block information with full transactions given the block id
+pub async fn get_block_with_txs(
+    context: RpcContext,
+    input: GetBlockInput,
+) -> Result<types::Block, GetBlockError> {
+    get_block(
+        context,
+        input.block_id,
+        types::BlockResponseScope::FullTransactions,
+    )
+    .await
+}
+
+/// Get block information given the block id
+async fn get_block(
+    context: RpcContext,
+    block_id: BlockId,
+    scope: types::BlockResponseScope,
+) -> Result<types::Block, GetBlockError> {
+    let block_id = match block_id {
+        BlockId::Pending => {
+            match context
+                .pending_data
+                .ok_or_else(|| anyhow!("Pending data not supported in this configuration"))?
+                .block()
+                .await
+            {
+                Some(block) => {
+                    return Ok(types::Block::from_sequencer_scoped(
+                        block.as_ref().clone().into(),
+                        scope,
+                    ))
+                }
+                None => StarknetBlocksBlockId::Latest,
+            }
+        }
+        BlockId::Hash(hash) => hash.into(),
+        BlockId::Number(number) => number.into(),
+        BlockId::Latest => StarknetBlocksBlockId::Latest,
+    };
+
+    let storage = context.storage.clone();
+    let span = tracing::Span::current();
+
+    tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+        let mut connection = storage
+            .connection()
+            .context("Opening database connection")?;
+
+        let transaction = connection
+            .transaction()
+            .context("Creating database transaction")?;
+
+        // Need to get the block status. This also tests that the block hash is valid.
+        let block = get_raw_block(&transaction, block_id)?;
+
+        let transactions = get_block_transactions(&transaction, block.number, scope)?;
+
+        Result::<types::Block, GetBlockError>::Ok(types::Block::from_raw(block, transactions))
```

Comment:
nit: is explicit typing necessary here? At first sight it seems like rustc should be able to deduct it on its own.

Reply:
lemme see..